File: dashboard/frontend/src/views/runtimes/nbody.js
// N-body runtime (client tier) — 2D softened gravity around a heavy centre,
// leapfrog-integrated: the real-time *approximation* of scenario 1. The precise
// counterpart is the server `nbody` (view_compute); this one is for orbits you
// can watch and perturb live. Pure module: create(spec) → { step, output }.

function rnd(a, b) { return a + Math.random() * (b - a); }

export function create(spec) {
  const bounds = spec?.bounds || [100, 100];
  const p0 = spec?.params || {};
  const [W, H] = bounds;
  const cx = W / 2; const cy = H / 2;
  const count = Math.min(600, Math.max(2, p0.count ?? 120));
  const scale = Math.min(W, H) * 0.42;

  // a disc of bodies on circular-ish orbits around a heavy centre
  const bodies = [{ x: cx, y: cy, vx: 0, vy: 0, m: p0.central_mass ?? 300 }];
  for (let i = 1; i < count; i++) {
    const r = Math.sqrt(rnd(0.05, 1)) * scale;
    const a = rnd(0, Math.PI * 2);
    const v = Math.sqrt((p0.g ?? 40) * bodies[0].m / Math.max(r, 1));
    bodies.push({
      x: cx + r * Math.cos(a), y: cy + r * Math.sin(a),
      vx: -v * Math.sin(a), vy: v * Math.cos(a), m: rnd(0.5, 1.5),
    });
  }

  function accel(out, g, soft2) {
    for (let i = 0; i < bodies.length; i++) { out[i * 2] = 0; out[i * 2 + 1] = 0; }
    for (let i = 0; i < bodies.length; i++) {
      for (let j = i + 1; j < bodies.length; j++) {
        const dx = bodies[j].x - bodies[i].x; const dy = bodies[j].y - bodies[i].y;
        const inv = Math.pow(dx * dx + dy * dy + soft2, -1.5) * g;
        out[i * 2] += inv * dx * bodies[j].m; out[i * 2 + 1] += inv * dy * bodies[j].m;
        out[j * 2] -= inv * dx * bodies[i].m; out[j * 2 + 1] -= inv * dy * bodies[i].m;
      }
    }
  }

  const acc = new Float64Array(count * 2);

  return {
    step(dt, params = {}) {
      const g = params.g ?? 40;
      const soft = params.softening ?? 2;
      const h = Math.min(dt, 1 / 30);
      accel(acc, g, soft * soft);
      for (let i = 0; i < bodies.length; i++) {
        const b = bodies[i];
        b.vx += acc[i * 2] * h * 0.5; b.vy += acc[i * 2 + 1] * h * 0.5;
        b.x += b.vx * h; b.y += b.vy * h;
      }
      accel(acc, g, soft * soft);
      for (let i = 0; i < bodies.length; i++) {
        bodies[i].vx += acc[i * 2] * h * 0.5; bodies[i].vy += acc[i * 2 + 1] * h * 0.5;
      }
    },
    output() {
      const positions = bodies.map((b) => [b.x, b.y, Math.hypot(b.vx, b.vy)]);
      let ke = 0;
      for (const b of bodies) ke += 0.5 * b.m * (b.vx * b.vx + b.vy * b.vy);
      return { bounds, positions, aggregates: { bodies: bodies.length, kinetic: ke } };
    },
  };
}

File: views/compute/nbody.py
"""Gravitational N-body — precise leapfrog (kick-drift-kick) integration.

Scenario 1 (cosmic objects & events). Softened Newtonian gravity; reports total
energy per frame as a conservation check (a fidelity signal). Positions stream
as a ``positions`` channel; the renderer draws them as points.
"""
from __future__ import annotations

from typing import Any, Dict

import numpy as np

_G = 1.0  # simulation units


class NBody:
    def __init__(self, params: Dict[str, Any]):
        n = int(min(2000, max(2, params.get("count", 200))))
        rng = np.random.default_rng(int(params.get("seed", 42)))
        self.softening = float(params.get("softening", 0.05))
        self.g = float(params.get("g", _G))
        # a rotating disc of bodies around a heavy centre → a mini galaxy
        radius = rng.uniform(0.2, 4.0, n) ** 0.5 * 3
        angle = rng.uniform(0, 2 * np.pi, n)
        self.pos = np.stack([radius * np.cos(angle), radius * np.sin(angle),
                             rng.normal(0, 0.15, n)], axis=1)
        self.mass = rng.uniform(0.5, 1.5, n)
        self.mass[0] = float(params.get("central_mass", 400.0))
        self.pos[0] = [0, 0, 0]
        # circular-ish initial velocities around the centre
        v = np.sqrt(self.g * self.mass[0] / (radius + self.softening))
        self.vel = np.stack([-v * np.sin(angle), v * np.cos(angle), np.zeros(n)], axis=1)
        self.vel[0] = [0, 0, 0]
        self._acc = self._accel()

    def _accel(self) -> np.ndarray:
        d = self.pos[None, :, :] - self.pos[:, None, :]           # (n,n,3)
        r2 = np.sum(d * d, axis=2) + self.softening ** 2
        inv = r2 ** -1.5
        np.fill_diagonal(inv, 0.0)
        return self.g * np.einsum("ij,ijk,j->ik", inv, d, self.mass)

    def step(self, dt: float) -> None:
        self.vel += 0.5 * self._acc * dt
        self.pos += self.vel * dt
        self._acc = self._accel()
        self.vel += 0.5 * self._acc * dt

    def frame(self) -> Dict[str, Any]:
        ke = float(0.5 * np.sum(self.mass * np.sum(self.vel ** 2, axis=1)))
        d = self.pos[None, :, :] - self.pos[:, None, :]
        r = np.sqrt(np.sum(d * d, axis=2) + self.softening ** 2)
        iu = np.triu_indices(len(self.mass), 1)
        pe = float(-self.g * np.sum(self.mass[iu[0]] * self.mass[iu[1]] / r[iu]))
        return {
            "positions": np.round(self.pos, 4).tolist(),
            "aggregates": {"bodies": int(len(self.mass)), "energy": round(ke + pe, 3)},
        }


def create(params: Dict[str, Any]) -> NBody:
    return NBody(params)
